File: tks/RepositoriesAdapters/src/test/kotlin/mappers/ResourceMapperTest.java
package mappers;

import data.AbstractResourceEntity;
import data.BookEntity;
import data.MagazineEntity;
import lombok.SneakyThrows;
import domain.model.Book;
import domain.model.Resource;
import domain.model.values.AccessionNumber;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Objects;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

public class ResourceMapperTest {

    ResourceMapper mapper;

    @BeforeEach
    public void init(){
        mapper = ResourceMapper.Companion.getINSTANCE();
    }

    @SneakyThrows
    @Test
    public void  BookToEntityTest(){
        Resource book = new Book(new AccessionNumber("EEEE-456"), "Diuna", "Frank Herbert");
        var entity = mapper.mapDomainObjectToEntity(book);
        assertNotNull(entity);
        assertTrue(entity instanceof BookEntity);
        assertEquals(Objects.requireNonNull(book.getAccessionNumber()).getValue(), entity.getAccessionNumber());
        assertEquals(book.getTitle(), entity.getTitle());
        assertNull(entity.getGuid());
        assertEquals(((Book)book).getAuthor(), ((BookEntity)entity).getAuthor());
    }

    @Test
    public void EntityToBookTest(){
        AbstractResourceEntity entity = new BookEntity("EEEE-456", "Diuna", "Frank Herbert");
        entity.setGuid(UUID.randomUUID());
        var book = mapper.mapEntityToDomainObject(entity);
        assertNotNull(book);
        assertTrue(book instanceof Book);
        assertEquals(entity.getAccessionNumber(), Objects.requireNonNull(book.getAccessionNumber()).getValue());
        assertEquals(entity.getTitle(), book.getTitle());
        assertEquals(((BookEntity)entity).getAuthor(), ((Book)book).getAuthor());
    }

    @SneakyThrows
    @Test
    public void BookToExistingEntityTest(){
        Resource book = new Book(new AccessionNumber("EEEE-456"), "Diuna", "Frank Herbert");
        AbstractResourceEntity entity = new BookEntity("EEEE-456", "Kolor magii", "Terry Pratchett");
        var guid = UUID.randomUUID();
        entity.setGuid(guid);
        mapper.mapDomainObjectToEntity(book, entity);
        assertEquals(Objects.requireNonNull(book.getAccessionNumber()).getValue(), entity.getAccessionNumber());
        assertEquals(book.getTitle(), entity.getTitle());
        assertEquals(((Book)book).getAuthor(), ((BookEntity)entity).getAuthor());
        assertEquals(guid, entity.getGuid());
    }

    @SneakyThrows
    @Test
    public void BookToExistingMagazineTest(){
        Resource book = new Book(new AccessionNumber("EEEE-456"), "Diuna", "Frank Herbert");
        AbstractResourceEntity entity = new MagazineEntity("EEEE-456", "Nature", "Nature Publishing Group");
        assertThrows(Exception.class, () -> mapper.mapDomainObjectToEntity(book, entity));
    }
}

File: tks/RepositoriesAdapters/src/test/kotlin/repository/mappers/ResourceMapperTest.kt
package repository.mappers

import domain.model.context.library.Book
import domain.model.context.library.Resource
import domain.model.values.AccessionNumber
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import repository.data.AbstractResourceEntity
import repository.data.BookEntity
import repository.data.MagazineEntity
import java.util.*

class ResourceMapperTest {
    private val mapper: ResourceMapper = ResourceMapper.INSTANCE

    @Test
    fun bookToEntityTest() {
        val book: Resource = Book(AccessionNumber("EEEE-456"), "Diuna", false, "Frank Herbert")
        val entity = mapper.mapDomainObjectToEntity(book)
        assertNotNull(entity!!)
        assertTrue(entity is BookEntity)
        assertEquals(book.accessionNumber?.value!!, entity.accessionNumber)
        assertEquals(book.title, entity.title)
        assertEquals(book.locked, entity.locked)
        assertNull(entity.guid)
        assertEquals((book as Book).author, (entity as BookEntity).author)
    }

    @Test
    fun nullableBookToEntity() {
        val entity = mapper.mapDomainObjectToEntity(null as Book?)
        assertNull(entity)
    }

    @Test
    fun entityToBookTest() {
        val entity: AbstractResourceEntity = BookEntity(UUID.randomUUID(), "EEEE-456", "Diuna", true, "Frank Herbert")
        val book = mapper.mapEntityToDomainObject(entity)
        assertNotNull(book!!)
        assertTrue(book is Book)
        assertEquals(entity.accessionNumber, book.accessionNumber?.value)
        assertEquals(entity.title, book.title)
        assertEquals(entity.locked, book.locked)
        assertEquals((entity as BookEntity).author, (book as Book).author)
    }

    @Test
    fun nullableEntityToBookTest() {
        assertNull(mapper.mapEntityToDomainObject(null as AbstractResourceEntity?))
    }

    @Test
    fun bookToExistingEntityTest() {
        val book: Resource = Book(AccessionNumber("EEEE-456"), "Diuna", true, "Frank Herbert")
        val guid = UUID.randomUUID()
        val entity: AbstractResourceEntity = BookEntity(guid, "EEEE-456", "Kolor magii", false, "Terry Pratchett")
        mapper.mapDomainObjectToEntity(book, entity)
        assertEquals(book.accessionNumber!!.value, entity.accessionNumber)
        assertEquals(book.title, entity.title)
        assertEquals(book.locked, entity.locked)
        assertEquals((book as Book).author, (entity as BookEntity).author)
        assertEquals(guid, entity.guid)
    }

    @Test
    fun nullableBookToExistingEntityTest(){
        val guid = UUID.randomUUID()
        val entity: AbstractResourceEntity = BookEntity(guid, "EEEE-456", "Kolor magii", true, "Terry Pratchett")
        mapper.mapDomainObjectToEntity(null as Resource?, entity)
        assertTrue(entity is BookEntity)
        assertEquals("EEEE-456", entity.accessionNumber)
        assertEquals("Kolor magii", entity.title)
        assertEquals(true, entity.locked)
        assertEquals("Terry Pratchett", (entity as BookEntity).author)
        assertEquals(guid, entity.guid)
    }

    @Test
    fun bookToExistingMagazineTest() {
        val book: Resource = Book(AccessionNumber("EEEE-456"), "Diuna", true, "Frank Herbert")
        val entity: AbstractResourceEntity = MagazineEntity(null, "EEEE-456", "Nature", true, "Nature Publishing Group")
        assertThrows<Exception> { mapper.mapDomainObjectToEntity(book, entity) }
    }
}